csharp/src/Google.Protobuf/Reflection/SingleFieldAccessor.cs

```diff
@@ -57,10 +57,11 @@ internal SingleFieldAccessor(PropertyInfo property, FieldDescriptor descriptor)
                 throw new ArgumentException("Not all required properties/methods available");
             }
             setValueDelegate = ReflectionUtil.CreateActionIMessageObject(property.GetSetMethod());
-            if (descriptor.File.Syntax == Syntax.Proto3)
+            if (descriptor.File.Syntax == Syntax.Proto3 && !descriptor.IsProto3Optional)
             {
-                hasDelegate = message => {
-                  throw new InvalidOperationException("HasValue is not implemented for proto3 fields");
+                hasDelegate = message =>
+                {
+                    throw new InvalidOperationException("HasValue is not implemented for non-optional proto3 fields");
```

Comment:
nit: s/implemented/supported/   -  "not implemented" sounds like something that should be implemented, but isn't yet. "Not supported" sounds like this is by design.  Leaving up to you whether you want to change this or not.

Reply:
Hmm. I really don't have strong feelings one way or another - I tend to lean towards "keep things as they are" when I'm on the fence.